blockchain/chain/src/store/chain_store.rs

```diff
@@ -261,7 +283,147 @@ where
         }
 
         // TODO would be better to get tipset with ChainStore cache.
-        tipset_by_height(self.blockstore(), lbr, ts, true)
+        self.tipset_by_height(lbr, ts, true)
+    }
+
+    /// Returns the tipset behind `tsk` at a given `height`.
+    /// If the given height is a null round:
+    /// - If `prev` is `true`, the tipset before the null round is returned.
+    /// - If `prev` is `false`, the tipset following the null round is returned.
+    ///
+    /// Returns `None` if the tipset provided was the tipset at the given height.
+    pub fn tipset_by_height(
+        &self,
+        height: ChainEpoch,
+        ts: &Tipset,
+        prev: bool,
+    ) -> Result<Option<Tipset>, Error> {
+        if height > ts.epoch() {
+            return Err(Error::Other(
+                "searching for tipset that has a height less than starting point".to_owned(),
+            ));
+        }
+        if height == ts.epoch() {
+            return Ok(None);
+        }
+        // TODO get tipset by height using cache instead of reloading tipsets
+        let mut ts_temp: Option<Tipset> = None;
+        loop {
+            let pts = if let Some(temp) = &ts_temp {
```

Comment:
Could you use the `tip_index.rs` here to act as the cache? If not what ts cache would that be?

Reply:
tip index is really broken right now and I'd rather not add any more usages until #814 is done. (I didn't change this code, just was moved)